File: 1647/greedy.py
from typing import *


class Solution:
    def minDeletions(self, s: str) -> int:
        freq_per_char = {}
        for c in s:
            if c not in freq_per_char:
                freq_per_char[c] = 0
            freq_per_char[c] += 1
        seen_freq = set([])
        delete_count = 0
        for freq in freq_per_char.values():
            while freq in seen_freq:
                delete_count += 1
                freq -= 1
            if freq > 0:
                seen_freq.add(freq)
        return delete_count


if __name__ == '__main__':
    sol = Solution()

    s, expected = "aab", 0
    assert sol.minDeletions(s) == expected

    s, expected = "aaabbbcc", 2
    assert sol.minDeletions(s) == expected

    s, expected = "ceabaacb", 2
    assert sol.minDeletions(s) == expected

    s, expected = "bbcebab", 2
    assert sol.minDeletions(s) == expected

File: 1647/greedy.rs
use std::collections::HashMap;
use std::collections::HashSet;

struct Solution;

impl Solution {
    pub fn min_deletions(s: String) -> i32 {
        let mut freq_per_char = HashMap::new();
        for c in s.chars() {
            let freq = freq_per_char.entry(c).or_insert(0);
            *freq += 1;
        }
        let mut seen_freq = HashSet::new();
        let mut delete_count = 0;
        for (_, orig_freq) in freq_per_char.iter() {
            let mut freq = *orig_freq;
            while seen_freq.contains(&freq) {
                delete_count += 1;
                freq -= 1;
            }
            if freq > 0 {
                seen_freq.insert(freq);
            }
        }
        return delete_count;
    }
}

fn main() {
    let (s, expected) = (String::from("aab"), 0);
    assert_eq!(Solution::min_deletions(s), expected);

    let (s, expected) = (String::from("aaabbbcc"), 2);
    assert_eq!(Solution::min_deletions(s), expected);

    let (s, expected) = (String::from("ceabaacb"), 2);
    assert_eq!(Solution::min_deletions(s), expected);

    let (s, expected) = (String::from("bbcebab"), 2);
    assert_eq!(Solution::min_deletions(s), expected);
}
